File: frontend/src/utils/authValidation.js
export const EMAIL_PATTERN = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
export const PASSWORD_PATTERN = /^(?=.*[A-Z])(?=.*\d)(?=.*[^A-Za-z0-9]).{8,}$/;

export const normalizeEmail = (email) => {
  const normalizedEmail = email.trim().toLowerCase();

  if (!EMAIL_PATTERN.test(normalizedEmail)) {
    const error = new Error('Please enter a valid email address.');
    error.code = 'auth/invalid-email';
    throw error;
  }

  return normalizedEmail;
};

export const validatePassword = (password) => {
  if (!PASSWORD_PATTERN.test(password)) {
    const error = new Error(
      'Password must contain at least 8 characters, one uppercase letter, one number, and one special symbol.'
    );
    error.code = 'auth/weak-password';
    throw error;
  }

  return password;
};

File: backend/app/services/auth_validation.py
"""Validation rules shared by backend authentication entry points."""

import re


EMAIL_PATTERN = re.compile(r"^[^\s@]+@[^\s@]+\.[^\s@]+$")
PASSWORD_PATTERN = re.compile(r"^(?=.*[A-Z])(?=.*\d)(?=.*[^A-Za-z0-9]).{8,}$")


def normalize_email(email: str) -> str:
    """Trim and normalize an email address, rejecting malformed input."""
    normalized = email.strip().lower()
    if not EMAIL_PATTERN.fullmatch(normalized):
        raise ValueError("Please enter a valid email address.")
    return normalized


def validate_password(password: str) -> str:
    """Validate the minimum password policy and return the original value."""
    if not PASSWORD_PATTERN.fullmatch(password):
        raise ValueError(
            "Password must contain at least 8 characters, one uppercase letter, "
            "one number, and one special symbol."
        )
    return password